### PR TITLE
hail2 -- reference genome support

### PR DESCRIPTION
**Pull request by jigold**
@tpoterba If this is too hard to rebase, I can make these same changes on top of your changes. I just wanted a record of what needed to be modified to make the genome reference functionality work.

### CODE REVIEW

File: python/hail/typ.py [hunk 1]
```diff
@@ -19,7 +20,7 @@ def __str__(self):
         return self.msg
 
 
-class Type(object):
+class Type(HistoryMixin):
```
Comment (tpoterba) [inline python/hail/typ.py]: This makes me nervous. Doesn't the stringified type for TVariant/Locus/Interval include the genome reference? If so, this is unnecessary.

Comment (jigold) [reply python/hail/typ.py]: It's true that I could just write the repr for TVariant/Locus... to print out the name of the reference. However, if we do this, then we will not have any information about how the reference was constructed in the history. And the repr for TVariant/Locus would no longer be a valid repr -- i.e. you can't construct a new GenomeReference object with just the name of the reference. If the repr is valid, you'll get something like `GenomeReference.GRCh37()` everywhere or even worse `GenomeReference(name="foo", contigs=["1", "2"...], lengths={"1": 100, "2": 200, ... })` making the history unusable.

My approach was to think of GenomeReference as similar to a HailContext. It's history should be set once on init and it's varid is set to the name of the reference. Here's an example:

```
# 2017-11-14T22:38:15.134288
# version: devel-b35882e

from hail import *

foo = (GenomeReference.from_file(
    file='src/test/resources/fake_ref_genome.json'))

(KeyTable.parallelize(rows=[{'v': Variant.parse(string='1:32:A:T', reference_genome=foo)}], schema=TStruct(names=['v'], types=[TVariant(reference_genome=foo, required=True)]))
 .export(output='/tmp/foo3.tsv'))
```

I could remove the mixin from singleton types, but Array, Dict, etc. would need it and I didn't see the harm in putting the mixin on Type.

File: python/hail/representation/genomeref.py [hunk 2]
```diff
@@ -66,21 +65,26 @@ def __init__(self, name, contigs, lengths, x_contigs=[], y_contigs=[], mt_contig
...
         self._y_contigs = y_contigs
         self._mt_contigs = mt_contigs
-        self._par = par
+        self._par = None
+        self._par_tuple = par
 
         super(GenomeReference, self).__init__()
 
+    def _set_history(self, history):
+        assert not self._history_was_set, "Cannot set history for GenomeReference more than once."
```
[the remark this answers is not in the document]
Comment (jigold) [reply python/hail/representation/genomeref.py]: It shouldn't. But I want to make sure it fails if it does.

[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/representation/genomeref.py]: ah, got it. Sounds good.

**Comment by jigold** [comment]
@tpoterba Can you take another look? If you want to close it and incorporate the changes into Hail2 later, I'm fine with that.

**Comment by tpoterba** [comment]
no, I think this is OK. I'll have to add the stuff at some point, might as well do it during a rebase!

**Pull request merged**